Crashing when trying to run FastAPI app on Python 3.14 RC and Debian (trixie)

Hi 👋 

I recently started trying out Python 3.14 on a FastAPI app running with granian. I've set up a simple app to demonstrate the crash: https://github.com/edgarrmondragon/--repro-granian-py314.

It works fine on my M2 MacBook:

```
uv run granian --interface=asgi main:app
```

but crashes on the official Python 3.14 image, regardless of the number of workers or whether I disable auto-reloading):

<details><summary>Details</summary>
<p>

```
[INFO] Starting granian (main PID: 1)
[INFO] Listening at: http://0.0.0.0:8000
[INFO] Spawning worker-1 with PID: 9
[INFO] Spawning worker-2 with PID: 11
[INFO] Spawning worker-3 with PID: 13
[INFO] Spawning worker-4 with PID: 15
[INFO] Started worker-1
[INFO] Started worker-3
[INFO] Started worker-2
[INFO] Started worker-2 runtime-1
[INFO] Started worker-3 runtime-1
[INFO] Started worker-1 runtime-1
web-1  |
thread '<unnamed>' panicked at src/workers.rs:163:58:
called `Result::unwrap()` on an `Err` value: Socket operation on non-socket (os error 88)
web-1  |
Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: PyObject_Vectorcall
   5: _PyEval_EvalFrameDefault
   6: <unknown>
   7: PyEval_EvalCode
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: Py_RunMain
  12: Py_BytesMain
  13: <unknown>
  14: __libc_start_main
  15: _start
stack backtrace:
   0:     0xffff934723f8 - <unknown>
   1:     0xffff933c6ef0 - <unknown>
   2:     0xffff93471f80 - <unknown>
   3:     0xffff93472298 - <unknown>
   4:     0xffff93471c5c - <unknown>
   5:     0xffff93497130 - <unknown>
   6:     0xffff934970a8 - <unknown>
   7:     0xffff93497d28 - <unknown>
   8:     0xffff92e5cf8c - <unknown>
   9:     0xffff92e5d2ec - <unknown>
  10:     0xffff931fdbd0 - <unknown>
  11:     0xffff93220bf8 - <unknown>
  12:     0xffff93265a7c - <unknown>
  13:     0xffff94b1f138 - PyObject_Vectorcall
  14:     0xffff94b35d6c - _PyEval_EvalFrameDefault
  15:     0xffff94b31c44 - <unknown>
  16:     0xffff94bf3044 - PyEval_EvalCode
  17:     0xffff94c52e60
thread '<unnamed>' panicked at src/workers.rs:163:58:
called `Result::unwrap()` on an `Err` value: Socket operation on non-socket (os error 88)
web-1  |
Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: PyObject_Vectorcall
   5: _PyEval_EvalFrameDefault
   6: <unknown>
   7: PyEval_EvalCode
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: Py_RunMain
  12: Py_BytesMain
  13: <unknown>
  14: __libc_start_main
  15: _start
 - <unknown>
stack backtrace:
  18:     0xffff94c4af94 - <unknown>
   19 :   0 :     0x ffff94c4ae30  - 0x<unknown>ffff934723f8
 - <unknown>
    20 : 1 :       0x ffff94be45940x - ffff933c6ef0Py_RunMain -
<unknown>
  21 :     2 :   0x ffff94bdb5c4  -  Py_BytesMain0x
ffff93471f80 - <unknown>
   22 :   3 :     0x ffff9486229c  - <unknown>0x
ffff93472298 - <unknown>
 23 :     4 :   0x ffff9486237c  -  __libc_start_main0x
ffff93471c5c - <unknown>
24:        50x: aaaae7320930  -  _start
 0xffff93497130 -  <unknown>
25:        6 :          0x ffff934970a8  -  <unknown>
 0x 0  -  <unknown>7
:     0xffff93497d28 - <unknown>
   8:     0xffff92e5cf8c - <unknown>
   9:     0xffff92e5d2ec - <unknown>
  10:     0xffff931fdbd0 - <unknown>
  11:     0xffff93220bf8 - <unknown>
  12:     0xffff93265a7c - <unknown>
  13:     0xffff94b1f138 - PyObject_Vectorcall
  14:     0xffff94b35d6c - _PyEval_EvalFrameDefault
  15:     0xffff94b31c44 - <unknown>
  16:     0xffff94bf3044 - PyEval_EvalCode
  17:     0xffff94c52e60 - <unknown>
  18:     0xffff94c4af94 - <unknown>
  19:     0xffff94c4ae30 - <unknown>
  20:     0xffff94be4594 - Py_RunMain
  21:     0xffff94bdb5c4 - Py_BytesMain
  22:     0xffff9486229c - <unknown>
  23:     0xffff9486237c - __libc_start_main
  24:     0xaaaae7320930 - _start
  25:                0x0 - <unknown>
web-1  |
thread '<unnamed>' panicked at src/workers.rs:163:58:
called `Result::unwrap()` on an `Err` value: Socket operation on non-socket (os error 88)
web-1  |
Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: PyObject_Vectorcall
   5: _PyEval_EvalFrameDefault
   6: <unknown>
   7: PyEval_EvalCode
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: Py_RunMain
  12: Py_BytesMain
  13: <unknown>
  14: __libc_start_main
  15: _start
stack backtrace:
   0:     0xffff934723f8 - <unknown>
   1:     0xffff933c6ef0 - <unknown>
   2:     0xffff93471f80 - <unknown>
   3:     0xffff93472298 - <unknown>
   4:     0xffff93471c5c - <unknown>
   5:     0xffff93497130 - <unknown>
   6:     0xffff934970a8 - <unknown>
   7:     0xffff93497d28 - <unknown>
   8:     0xffff92e5cf8c - <unknown>
   9:     0xffff92e5d2ec - <unknown>
  10:     0xffff931fdbd0 - <unknown>
  11:     0xffff93220bf8 - <unknown>
  12:     0xffff93265a7c - <unknown>
  13:     0xffff94b1f138 - PyObject_Vectorcall
  14:     0xffff94b35d6c - _PyEval_EvalFrameDefault
  15:     0xffff94b31c44 - <unknown>
  16:     0xffff94bf3044 - PyEval_EvalCode
  17:     0xffff94c52e60 - <unknown>
  18:     0xffff94c4af94 - <unknown>
  19:     0xffff94c4ae30 - <unknown>
  20:     0xffff94be4594 - Py_RunMain
  21:     0xffff94bdb5c4 - Py_BytesMain
  22:     0xffff9486229c - <unknown>
  23:     0xffff9486237c - __libc_start_main
  24:     0xaaaae7320930 - _start
  25:                0x0 - <unknown>
[INFO] Started worker-4
[INFO] Started worker-4 runtime-1
web-1  |
thread '<unnamed>' panicked at src/workers.rs:163:58:
called `Result::unwrap()` on an `Err` value: Socket operation on non-socket (os error 88)
web-1  |
Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: PyObject_Vectorcall
   5: _PyEval_EvalFrameDefault
   6: <unknown>
   7: PyEval_EvalCode
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: Py_RunMain
  12: Py_BytesMain
  13: <unknown>
  14: __libc_start_main
  15: _start
stack backtrace:
   0:     0xffff934723f8 - <unknown>
   1:     0xffff933c6ef0 - <unknown>
   2:     0xffff93471f80 - <unknown>
   3:     0xffff93472298 - <unknown>
   4:     0xffff93471c5c - <unknown>
   5:     0xffff93497130 - <unknown>
   6:     0xffff934970a8 - <unknown>
   7:     0xffff93497d28 - <unknown>
   8:     0xffff92e5cf8c - <unknown>
   9:     0xffff92e5d2ec - <unknown>
  10:     0xffff931fdbd0 - <unknown>
  11:     0xffff93220bf8 - <unknown>
  12:     0xffff93265a7c - <unknown>
  13:     0xffff94b1f138 - PyObject_Vectorcall
  14:     0xffff94b35d6c - _PyEval_EvalFrameDefault
  15:     0xffff94b31c44 - <unknown>
  16:     0xffff94bf3044 - PyEval_EvalCode
  17:     0xffff94c52e60 - <unknown>
  18:     0xffff94c4af94 - <unknown>
  19:     0xffff94c4ae30 - <unknown>
  20:     0xffff94be4594 - Py_RunMain
  21:     0xffff94bdb5c4 - Py_BytesMain
  22:     0xffff9486229c - <unknown>
  23:     0xffff9486237c - __libc_start_main
  24:     0xaaaae7320930 - _start
  25:                0x0 - <unknown>
[ERROR] Unexpected exit from worker-1
[ERROR] Unexpected exit from worker-2
[ERROR] Unexpected exit from worker-3
[ERROR] Unexpected exit from worker-4
[INFO] Shutting down granian
```

</p>
</details> 

It also works fine under 3.13.

Here's a repro, with Starlette instead of FastAPI:

```bash
mkdir /tmp/granian-test
cd /tmp/granian-test
uv venv --python 3.14.0rc2
uv pip install starlette granian
cat >main.py <<EOF
from starlette.applications import Starlette
from starlette.responses import HTMLResponse
from starlette.routing import Route
async def homepage(request):
    return HTMLResponse('<h1>Hello, world!</h1>')
app = Starlette(debug=True, routes=[Route('/', homepage)])
EOF
RUST_BACKTRACE=1 uv run granian main:app
```

Output:

```text
[INFO] Starting granian (main PID: 23513)
[INFO] Listening at: http://127.0.0.1:8000
[INFO] Spawning worker-1 with PID: 23516
[INFO] Started worker-1
[INFO] Started worker-1 runtime-1

thread '<unnamed>' panicked at src/workers.rs:163:58:
called `Result::unwrap()` on an `Err` value: Socket operation on non-socket (os error 88)

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: _TAIL_CALL_CALL.llvm.7457709014681231701
   5: _PyEval_Vector
   6: PyEval_EvalCode
   7: run_mod.llvm.2886862816012006527
   8: _PyRun_SimpleStringFlagsWithName
   9: Py_RunMain
  10: pymain_main.llvm.13044120238892403703
  11: main
  12: <unknown>
  13: __libc_start_main
  14: _start
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[ERROR] Unexpected exit from worker-1
[INFO] Shutting down granian
```

Hope this helps!

--- a/granian/server/mp.py
+++ b/granian/server/mp.py
@@ -33,6 +33,14 @@ multiprocessing.allow_connection_pickling()
 class WorkerProcess(AbstractWorker):
     _idl = 'PID'
 
+    def __init__(self, parent, idx, target, args):
+        # NOTE: Python 3.14 defaults mp spawn method to 'forkserver' on Linux,
+        #       which doesn't really play well with shared sockets.
+        self._spawn_method = multiprocessing.get_start_method()
+        if self._spawn_method not in {'fork', 'spawn'}:
+            self._spawn_method = 'spawn'
+        super().__init__(parent, idx, target, args)
+
     @staticmethod
     def wrap_target(target):
         @wraps(target)
@@ -68,7 +76,9 @@ class WorkerProcess(AbstractWorker):
         return wrapped
 
     def _spawn(self, target, args):
-        self.inner = multiprocessing.get_context().Process(name='granian-worker', target=target, args=args)
+        self.inner = multiprocessing.get_context(self._spawn_method).Process(
+            name='granian-worker', target=target, args=args
+        )
 
     def _id(self):
         return self.inner.pid

